### PR TITLE
Avoid assigning an XID to DTX_CONTEXT_QE_AUTO_COMMIT_IMPLICIT queries.

### PR DESCRIPTION
**Pull request by hlinnaka**
This avoids a lot of overhead for short read-only queries. We still don't
do lazy assignment for most transactions, like PostgreSQL 8.3 does, but
this is a step in the right direction.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Passes installcheck-good on my laptop. The reduced overhead can be shown with this benchmark:

```
$ cat a.sql
SELECT * FROM foo WHERE id = 1;

BEFORE:
$ ~/gpdb.master/bin/pgbench  -f a.sql -n postgres -T10 
transaction type: Custom query
scaling factor: 1
query mode: simple
number of clients: 1
number of threads: 1
duration: 10 s
number of transactions actually processed: 1952
tps = 195.198575 (including connections establishing)
tps = 195.295441 (excluding connections establishing)


AFTER:
$ ~/gpdb.master/bin/pgbench  -f a.sql -n postgres -T10 
transaction type: Custom query
scaling factor: 1
query mode: simple
number of clients: 1
number of threads: 1
duration: 10 s
number of transactions actually processed: 4038
tps = 403.695201 (including connections establishing)
tps = 403.879807 (excluding connections establishing)

```

This is still a far cry from PostgreSQL 8.3, which does over 6000 tps in the same test, but it's a big improvement over the status quo.

**Comment by yydzero** [comment]
Great improvement!

File: src/backend/access/transam/xact.c [hunk 1]
```diff
@@ -3056,19 +3055,14 @@ StartTransaction(void)
 				 *
 				 * Generate a new transaction id
 				 */
-				s->transactionId = GetNewTransactionId(false, true);
+				AssignTransactionId(s);
```
Comment (hlinnaka) [inline src/backend/access/transam/xact.c]: AssignTransactionId() calls XactLockTableInsert().

**Comment by hlinnaka** [comment]
I interpreted the "Great improvement!" as a +1, and committed.

**Pull request merged**